src/gradualizer.erl

```diff
@@ -83,23 +78,234 @@ type_check_dir(Dir, Opts) ->
+format_error_(LINE, {mfa, M, F, A, ResTy}, {type_mismatch, FunTy}) ->
+    io_lib:format("The mfa ~p:~p/~p on line ~p is expected to have type: ~n~s~n"
+        "but has type : ~n"
+        "~s~n",
+        [M, F, A, LINE, typelib:pp_type(ResTy),
+        typelib:pp_intersection_type(FunTy)]);
+format_error_(LINE, {result, Func, FunResTy}, {type_mismatch, ResTy}) ->
+    io_lib:format("The function ~s on line ~p is expected to return ~s"
+        " but it returns ~s~n",
+        [erl_pp:expr(Func), LINE, typelib:pp_type(ResTy),
+            typelib:pp_type(FunResTy)]);
+format_error_(LINE, {function, Func, FunTy}, not_function_type) ->
+    io_lib:format("Expected function ~s on line ~p to have a function type,~n"
+        "but it has the following type:~n~s~n",
+        [erl_pp:expr(Func), LINE, typelib:pp_type(FunTy)]);
+format_error_(LINE, {call, Func}, {intersection_mismatch, FunTy}) ->
+    io_lib:format("None of the types of the function ~s at line ~p matches the "
+        "call site. Here's the types of the function:~n~s~n",
+        [erl_pp:expr(Func), LINE, typelib:pp_intersection_type(FunTy)]);
+format_error_(LINE, {bool_op, BoolOp, Ty}, not_boolean) ->
+    io_lib:format("The operator ~p on line ~p is given a non-boolean argument "
+        "of type ~s~n", [BoolOp, LINE, typelib:pp_type(Ty)]);
+format_error_(LINE, {rel_op, RelOp, Ty1, Ty2}, incompatible) ->
+    io_lib:format("The operator ~p on line ~p requires arguments of "
+        "compatible types.~nHowever, it has arguments "
+        "of type ~s and ~s~n",
+        [RelOp, LINE, typelib:pp_type(Ty1), typelib:pp_type(Ty2)]);
+format_error_(LINE, {arith_op, ArithOp, Ty}, non_numeric) ->
+    io_lib:format("The operator ~p on line ~p is given a non-numeric argument "
+        "of type ~s~n",
+        [ArithOp, LINE, typelib:pp_type(Ty)]);
+format_error_(LINE, {int_op, IntOp, Ty}, non_integer) ->
+    io_lib:format("The operator ~p on line ~p is given a non-integer argument "
+        "of type ~s~n",
+        [IntOp, LINE, typelib:pp_type(Ty)]);
+format_error_(LINE, unary_minus, {non_numeric_exp, Ty}) ->
+    io_lib:format("The negated expression on line ~p is expected to have a "
+        "non-numeric type:~n~s~n",
+        [LINE, typelib:pp_type(Ty)]);
+format_error_(LINE, {unary_minus, Ty}, non_numeric) ->
+    io_lib:format("The negated expression on line ~p has a non-numeric argument"
+        " of type:~n~s~n",
+        [LINE, typelib:pp_type(Ty)]);
+format_error_(LINE, {logic_op, LogicOp, Ty}, non_boolean) ->
+    io_lib:format("The operator ~p on line ~p is given a non-boolean argument "
+        "of type ~s~n",
+        [LogicOp, LINE, typelib:pp_type(Ty)]);
+format_error_(LINE, {logic_op, LogicOp}, {non_boolean_exp, Ty}) ->
+    io_lib:format("The operator ~p on line ~p is used in a context where it is "
+        "required to have type ~s~n",
+        [LogicOp, LINE, typelib:pp_type(Ty)]);
+format_error_(LINE, {logic_op, LogicOp, Ty1, Ty2}, incompatible) ->
```

Comment:
Apart from the line number element, is there a rule of thumb for what the elements are used for? Here, they are two values which tagged with a descriptive atoms instead of a flat structure of variable number of elements. I don't see how this is better.